### PR TITLE
Unit tests and some bug fixes

### LINKED ISSUE #33: Make beprof working with numpy 1.8
Numpy 1.8 is shipped by default in Ubuntu 14.04 LTS, we should support it

### PR DESCRIPTION
**Pull request by antnieszka**
- Implements #99 
- Improves #19 
- Fixes #33, #117, #111, #116, #113, #118, #115

### CODE REVIEW

File: beprof/curve.py [hunk 1]
```diff
@@ -113,7 +113,8 @@ def change_domain(self, domain):
             raise ValueError('in change_domain():' 'the old domain does not include the new one')
 
         y = np.interp(domain, self.x, self.y)
-        obj = self.__class__(np.stack((domain, y), axis=1), **self.__dict__['metadata'])
+        # np.dstack(...)[0] is used to extract nested array (previously used np.stack which behaved different)
+        obj = self.__class__(np.dstack((domain, y))[0], **self.__dict__['metadata'])
```
Comment (grzanka) [inline beprof/curve.py:160]: Add a comment why dstack is used here ? Why is it needed, what is it doing ?

Comment (antnieszka) [reply beprof/curve.py:160]: Like "stack is available from numpy 1.10 and to preserve compatibility we use older function which does the same thing as np.stack((arrays), axis=1)" ?

**Comment by antnieszka** [comment]
I'm writing missing test for normalize and this came out:
```py
>>> a
Profile([[ 1,  1],
       [ 2, 20],
       [ 3, 40]])
>>> a.normalize(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ant6/.virtualenvs/in_beprof3/lib/python3.4/site-packages/beprof/profile.py", line 150, in normalize
    self.y /= ave
TypeError: ufunc 'true_divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
>>> ave = np.average(a.y[np.fabs(a.x) <= 0])
>>> ave
nan
>>> a /= ave
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ufunc 'true_divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
```
Our method looks like this:
```py
    def normalize(self, dt):
        """
        Normalize to 1 over [-dt, +dt] area
        :param dt:
        :return:
        """
        logger.info('Running {0}.normalize(dt={1})'.format(self.__class__, dt))
        try:
            ave = np.average(self.y[np.fabs(self.x) <= dt])
        except RuntimeWarning as e:
            logger.error('in normalize(). self class is {0}, dt={1}'.format(self.__class__, dt))
            raise Exception("Scaling factor error:\n" + str(e))
        self.y /= ave  # <-- extra try-except?
```
So it's not idiot-proof enough.

**Comment by antnieszka** [comment]
It's something with division...
```py
>>> a
Profile([[ 1,  1],
       [ 2, 20],
       [ 3, 40]])
>>> ave = np.average(a.y[np.fabs(a.x) <= 10])
>>> ave
20.333333333333332
>>> a.y /= ave
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ufunc 'true_divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
```
```
(in_beprof3){16-12-05 22:18} ~ ant6% ➜ python -V
Python 3.4.5
(in_beprof3){16-12-05 22:18} ~ ant6% ➜ pip freeze
beprof==0.post0.dev1801480862596
numpy==1.11.2
```

**Comment by antnieszka** [comment]
@grzanka that's sad:
```
>>> import numpy as np
>>> a = np.array([1,2,3])
>>> b = 2.0
>>> a /= b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ufunc 'true_divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
```
More here: https://github.com/nipy/dipy/issues/730/ and https://github.com/numpy/numpy/issues/6464
And this works:
```
>>> a = np.array([1.0, 2.0, 3.0])
>>> a /= 2.0
>>> a
array([ 0.5,  1. ,  1.5])
```
when `a` is filled with floats...

**Comment by antnieszka** [comment]
In other words - Profile.y data has to be floats or it errors...

File: beprof/curve.py (hunk 1 above)
Comment (grzanka) [reply beprof/curve.py:160]: Not really, I had in mind a description for a fellow programmer who takes over the project and is trying to figure out what is going on in these few lines of code, sth like:

```We need to ... domain and y because .... We pass it as argument to ... because it takes .... Dstast is ... and this is the reason why we use [0]```

**Comment by grzanka** [comment]
Typically we will work with floats. But you are right - we should make a separate issue for Y data with ints and methods which use division.
There should be better user-friendy error handling and we can also provide some optional casting to float in such method.

File: tests/test_profile.py [hunk 2]
```diff
@@ -53,48 +54,48 @@ class TestProfile(TestCase):
...
+        self.p = Profile([[0.0, 5.0], [0.1, 10.0], [0.2, 20.0], [0.3, 10.0]])
 
     def test_x_at_y(self):
-        p = Profile([[0.0, 5.0], [0.1, 10.0], [0.2, 20.0], [0.3, 10.0]])
         # points outside profile range [5.0, 20.0]
-        assert np.isnan(p.x_at_y(-1.0))
-        assert np.isnan(p.x_at_y(4.9999999))
-        assert np.isnan(p.x_at_y(20.0000001))
-        assert np.isnan(p.x_at_y(7.5, reverse=True))
+        assert np.isnan(self.p.x_at_y(-1.0))
```
Comment (grzanka) [inline tests/test_profile.py]: Why `assert` instead of `self.assertTrue` ?

Comment (antnieszka) [reply tests/test_profile.py]: Probably inspired by numpy docs... I'll change it in the upcoming commit.

**Comment by antnieszka** [comment]
@grzanka I think, we can now change required version of numpy in setup.py to 1.8.0?

**Comment by grzanka** [comment]
Yes, change it please

File: beprof/curve.py [hunk 3]
```diff
@@ -113,7 +113,13 @@ def change_domain(self, domain):
             raise ValueError('in change_domain():' 'the old domain does not include the new one')
 
         y = np.interp(domain, self.x, self.y)
-        obj = self.__class__(np.stack((domain, y), axis=1), **self.__dict__['metadata'])
+        # We need to join together domain and y because we are recreating 2 dimensional Curve object
+        # (we pass it as argument to self.__class__ to do so and it takes 2 or 3 dimensional arrays as argument.
```
Comment (grzanka) [inline beprof/curve.py]: We should have a discussion on curve with 3-D data. Check if it is implemented correctly, test methods etc... - separate issue.

File: beprof/profile.py [hunk 4]
```diff
@@ -144,11 +146,14 @@ def normalize(self, dt):
         except RuntimeWarning as e:
             logger.error('in normalize(). self class is {0}, dt={1}'.format(self.__class__, dt))
             raise Exception("Scaling factor error:\n" + str(e))
-        self.y /= ave
+        self.y = self.y / ave
```
Comment (grzanka) [inline beprof/profile.py]: Why this change was made ?

File: beprof/profile.py [hunk 5]
```diff
@@ -144,11 +146,14 @@ def normalize(self, dt):
...
 
     def __str__(self):
         logger.info('Running {0}.__str__'.format(self.__class__))
-        ret = super().__str__()
+        # todo: investigate why super().__str__() errors in python 2.7
+        #       after change of implementation of __new__
+        ret = curve.Curve.__str__(self)
+        # ret = super(curve.Curve).__str__()
```
Comment (grzanka) [inline beprof/profile.py]: Why did you added second commented line ?

File: tests/test_profile.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        p1.normalize(2)
+        self.assertTrue(np.array_equal(p1.y, [0, 1, 3]))
+
+        p1 = Profile([[1, 1], [2, 20], [3, 40]])
+        p1.normalize(100)
+        self.assertTrue(np.array_equal(p1.y, [0, 0, 1]))
+
+        # case - less or equal to 0
+        p1 = Profile([[1, 1], [2, 20], [3, 40]])
+        p1.normalize(-1)
```
Comment (grzanka) [inline tests/test_profile.py]: Does it make sense to normalize this way ?

File: beprof/profile.py (hunk 5 above)
Comment (antnieszka) [reply beprof/profile.py]: It was the previous version left to discussion, but since we're using git - it's pointless, yes.

File: tests/test_profile.py (hunk 6 above)
Comment (antnieszka) [reply tests/test_profile.py]: No, but we probably should inform user about it. Maybe `raise ValueError()`?

File: beprof/profile.py (hunk 4 above)
Comment (antnieszka) [reply beprof/profile.py]: Fixing error mentioned above in comments:
```py
>>> c = np.array([1,2,3])
>>> c
array([1, 2, 3])
>>> c /= 2.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ufunc 'divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
>>> c
array([1, 2, 3])
>>> c =  c / 2.
>>> c
array([ 0.5,  1. ,  1.5])
```

File: tests/test_profile.py (hunk 6 above)
Comment (grzanka) [reply tests/test_profile.py]: Good idea. Such exception is OK, but add also some message in it, guiding user what type of input is expected.

File: beprof/profile.py (hunk 4 above)
Comment (antnieszka) [reply beprof/profile.py]: When we have only integers - should I roll back this one?

File: beprof/profile.py (hunk 5 above)
Comment (grzanka) [reply beprof/profile.py]: Remove it, also "investigate TODO" can directly go as new issue, instead of dirty comment.

File: beprof/curve.py (hunk 3 above)
Comment (antnieszka) [reply beprof/curve.py]: Done: #114

File: beprof/profile.py (hunk 5 above)
Comment (antnieszka) [reply beprof/profile.py]: It already has issue: #99

File: tests/test_profile.py (hunk 6 above)
Comment (antnieszka) [reply tests/test_profile.py]: Probably "Expected positive input" would be enough?

File: beprof/profile.py (hunk 4 above)
Comment (grzanka) [reply beprof/profile.py]: You can keep this change, but add info in docstring what happens if you have array of integers, explaining what if you:
 - normalize with integer factor
 - normalize with floating-point factor

But before proceeding further: consider these two snippets: `self.y /= ave` and `self.y = self.y / ave` - which one requires additional memory allocation ?

What if `self.y` is large array or floats (say 1e6) and we normalize by factor 0.5. Which one of abovementioned snippets will run faster ?

Reading https://docs.scipy.org/doc/numpy/reference/internals.html might help.

Comment (antnieszka) [reply beprof/profile.py]: Ok, I have to ideas for `rescale` and `normalize` divisions:
```py
    def rescale(self, factor=1.0, allow_cast=True):
        if allow_cast:  # will work in 99.9% cases
            try:
                self.y /= factor
            # there is this case, where "TypeError: ufunc 'true_divide' output (typecode 'd') could not be coerced
            # to provided output parameter (typecode 'l') according to the casting rule ''same_kind''" i thrown
            except TypeError as e: 
                logger.info("Division in place is impossible, casting...\n%s" % e)
                self.y = self.y / factor
        elif not np.issubdtype(self.y.dtype.type, type(factor)):
            raise TypeError("Type mismatch - array dtype: %s, factor type: %s\nallow_cast flag set to True should help"
                            % (self.y.dtype.type, type(factor)))
        else:
            try:
                self.y /= factor
            except TypeError as e:
                raise TypeError("%s\nallow_cast flag set to True should help" % e)
```
and simpler(?) one:
```py
def rescale(self, factor=1.0, allow_cast=True):
    try:
        self.y /= factor
    except TypeError as e: 
        logger.info("Division in place is impossible.\n%s" % e)
        if allow_cast:
            self.y = self.y / factor
        else:
            raise TypeError("%s\nallow_cast flag set to True should help" % e)
```

Comment (grzanka) [reply beprof/profile.py]: Second one better, but please add also proper docstring with doctests (see i.e. change_domain and rebinned methods).

Read also this: http://www.ianbicking.org/blog/2007/09/re-raising-exceptions.html
https://blog.ionelmc.ro/2014/08/03/the-most-underrated-feature-in-python-3/

Comment (antnieszka) [reply beprof/profile.py]: Sure, the second came to my mind when I was writing this reply, so...

What about the `div zero` case? Should we somehow catch it, or just let it error or whatever?

Comment (grzanka) [reply beprof/profile.py]: Print nice error message and let numpy exception be thrown

Comment (antnieszka) [reply beprof/profile.py]: Well, the above is not yet ready...
```py
>a
Curve([[ 0,  0],
       [ 5,  5],
       [10, 10]])
>a.rescale(3.7)
>a
Curve([[ 0,  0],
       [ 5,  1],
       [10,  2]])
>a.rescale(22)
>a
Curve([[ 0,  0],
       [ 5,  0],
       [10,  0]])
```

File: beprof/curve.py [hunk 7]
```diff
@@ -75,8 +78,21 @@ def y(self):
...
+        :return:
+        """
+        try:
+            self.y /= factor
+        except TypeError as e:
+            logger.error("Division in place is impossible.\n%s" % e)
+            if allow_cast:
+                self.y = self.y / factor
+            else:
+                logger.error("allow_cast flag set to True should help")
```
Comment (grzanka) [inline beprof/curve.py:115]: Does code execution continue or stops with some exception being rethrown ?

Comment (antnieszka) [reply beprof/curve.py:115]: It stops, we could go for re-raise at the end - after calculating with ` self.y = self.y / factor` or ignore TypeErrors thrown by  `self.y /= factor` or... just log it?
```py
        try:
            self.y /= factor
        except TypeError as e:
            logger.error("Division in place is impossible")
            if allow_cast:
                self.y = self.y / factor
            else:
                logger.error("allow_cast flag set to True should help")
            raise e
```

Comment (antnieszka) [reply beprof/curve.py:115]: Ignore error case (clean, but...):
```
>>> a
Curve([[1, 5],
       [2, 6],
       [3, 7],
       [4, 8]])
>>> a.rescale(2)
Division in place is impossible
>>> a
Curve([[1, 2],
       [2, 3],
       [3, 3],
       [4, 4]])
>>> a.rescale(2, allow_cast=False)
Division in place is impossible
allow_cast flag set to True should help
>>> a
Curve([[1, 2],
       [2, 3],
       [3, 3],
       [4, 4]])
```
Works like this for code below:
```py
    def rescale(self, factor=1.0, allow_cast=True):
        try:
            self.y /= factor
        except TypeError as e:
            logger.error("Division in place is impossible")
            if allow_cast:
                self.y = self.y / factor
            else:
                logger.error("allow_cast flag set to True should help")
```

Comment (antnieszka) [reply beprof/curve.py:115]: The thing is - at least for me if something throws exception - it probably did not finish it's job (and possibly rolled back). So raising exception after successful division is... could be misinterpreted?

File: beprof/profile.py [hunk 8]
```diff
@@ -132,23 +132,36 @@ def fwhm(self):
...
+            raise Exception("Scaling factor error: {0}".format(e))
+        try:
+            self.y /= ave
+        except TypeError as e:
+            logger.warning("Division in place is impossible: {0}".format(e))
+            if allow_cast:
+                self.y = self.y / ave
+            else:
+                logger.error("Division in place impossible - allow_cast flag set to True should help")
+                raise e
```
Comment (grzanka) [inline beprof/profile.py]: Is it the correct way of re-throwing an exception ?
See i.e. http://blog.bstpierre.org/python-exception-handling-cleanup-and-reraise

Comment (antnieszka) [reply beprof/profile.py]: Not really. Another try-catch would be nice and reporting original exception - like `reraiser4()`?

Edit: some of this examples are no longer supported e.g. in Python3.4

Comment (antnieszka) [reply beprof/profile.py]: For:
```py
        try:
            self.y /= factor
        except TypeError as e:
            # (_, _, traceback) = sys.exc_info()
            logger.warning("Division in place is impossible: {0}".format(e))
            if allow_cast:
                self.y = self.y / factor
            else:
                logger.error("allow_cast flag set to True should help")
            raise
```
result:
```
>>> a.rescale(0, False)
Division in place is impossible: ufunc 'true_divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
allow_cast flag set to True should help
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/home/ant6/PycharmProjects/beprof/beprof/curve.py", line 111, in rescale
    self.y /= factor
TypeError: ufunc 'true_divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
```
result2
```
>>> a.rescale(0, True)
Division in place is impossible: ufunc 'true_divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
/home/ant6/PycharmProjects/beprof/beprof/curve.py:116: RuntimeWarning: divide by zero encountered in true_divide
  self.y = self.y / factor
/home/ant6/PycharmProjects/beprof/beprof/curve.py:116: RuntimeWarning: invalid value encountered in true_divide
  self.y = self.y / factor
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/home/ant6/PycharmProjects/beprof/beprof/curve.py", line 111, in rescale
    self.y /= factor
TypeError: ufunc 'true_divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
```
Both exceptions are reported

**Comment by antnieszka** [comment]
@grzanka ready for re-review

**Review by grzanka — APPROVED** [review]
